File: packages/audio-context/src-native/android/app/src/main/java/org/nativescript/audiocontext/examples/VisualizerView.java
package org.nativescript.audiocontext.examples;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

public class VisualizerView extends View {
    private float[] data;
    private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public VisualizerView(Context c) { super(c); init(); }
    public VisualizerView(Context c, AttributeSet a) { super(c, a); init(); }
    public VisualizerView(Context c, AttributeSet a, int s) { super(c, a, s); init(); }

    private void init() {
        paint.setColor(Color.rgb(50, 200, 120));
    }

    public void setData(float[] d) {
        this.data = d;
        invalidate();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        int w = getWidth();
        int h = getHeight();
        canvas.drawColor(Color.rgb(8, 12, 26));
        if (data == null || data.length == 0) return;

        int barCount = Math.min(64, data.length);
        int binsPerBar = Math.max(1, data.length / barCount);
        float barWidth = (float) w / (float) barCount;

        final float minDb = -100f;
        final float maxDb = -30f;
        final float range = Math.max(1.0f, maxDb - minDb);

        for (int i = 0; i < barCount; ++i) {
            float sum = 0f;
            int base = i * binsPerBar;
            int end = Math.min(data.length, base + binsPerBar);
            for (int j = base; j < end; ++j) sum += data[j];
            float avg = sum / (end - base);
            float norm = (avg - minDb) / range;
            if (norm < 0f) norm = 0f;
            if (norm > 1f) norm = 1f;
            float barH = norm * h;
            float left = i * barWidth + 2f;
            float right = left + barWidth - 4f;
            canvas.drawRect(left, h - barH, right, h, paint);
        }
    }
}

File: packages/audio-context/src-native/android/app/src/main/java/org/nativescript/audiocontextdemo/VisualizerView.kt
package org.nativescript.audiocontextdemo

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import java.nio.ByteBuffer
import androidx.core.graphics.toColorInt

class VisualizerView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {
    private val paint = Paint()
    private var analyser: org.nativescript.audiocontext.AnalyserNode? = null
    private var buf: ByteBuffer? = null
		private var arr: ByteArray? = null
    private val frameDelayMs = 16L
    private val updater = object : Runnable {
        override fun run() {
            invalidate()
            postDelayed(this, frameDelayMs)
        }
    }

    fun setAnalyser(a: org.nativescript.audiocontext.AnalyserNode?) {
        analyser = a
        // allocate buffer sized to frequency bin count
        buf = null
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        post(updater)
    }

    override fun onDetachedFromWindow() {
        removeCallbacks(updater)
        super.onDetachedFromWindow()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val a = analyser ?: return
        val binCount = try { a.frequencyBinCount
				} catch (_: Throwable) { 64 }
        if (binCount <= 0) return

        if (buf == null || buf!!.capacity() < binCount) {
            buf = ByteBuffer.allocateDirect(binCount)
        }
        val b = buf!!
        b.position(0)
        try {
            a.getByteFrequencyData(b)
        } catch (e: Throwable) {
            // fallthrough
        }

        // read bytes
				if (arr == null){
					arr = ByteArray(binCount)
				}
				val arr = arr!!

        for (i in 0 until binCount) arr[i] = b.get(i)

        val w = width.toFloat()
        val h = height.toFloat()
        canvas.drawColor("#071026".toColorInt())

        val barCount = 64
        val step = 1.coerceAtLeast(binCount / barCount)
        val barWidth = (w / barCount) * 0.9f
        var x = 0f
        for (i in 0 until barCount) {
            val vi = (arr[i * step].toInt() and 0xff)
            val v = vi / 255f
            val barHeight = v * h
            val r = (200 * v + 55).toInt().coerceIn(0,255)
            val g = (120 * (1 - v) + 80).toInt().coerceIn(0,255)
            paint.color = Color.rgb(r, g, 120)
            canvas.drawRect(x, h - barHeight, x + barWidth, h, paint)
            x += barWidth + 2
        }
    }
}
